### PR TITLE
Add more property to GroupBroadcastMessage

### PR DESCRIPTION
**Pull request by vicancy**
Provide the ability for group broadcast to exclude users, and also add "senderId" property to group broadcast for client pub-sub

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs [hunk 1]
```diff
@@ -869,6 +871,8 @@ private static GroupBroadcastDataMessage CreateGroupBroadcastDataMessage(ref Mes
             {
                 result.ReadExtensionMembers(ref reader);
             }
+            result.ExcludedUserList = ReadStringArray(ref reader, "excludedUserList");
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs:874]: check `if (arrayLength >= 7)`

File: test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs [hunk 2]
```diff
@@ -497,15 +547,6 @@ public static IEnumerable<object[]> TestParseOldData
...
                 binary: "lBGmY29ubjExpmdyb3VwMoEBzQTS"),
-            new ProtocolTestData(
-                name: "GroupBroadcastWithTracingId",
-                message: new GroupBroadcastDataMessage("group3",
-                    new Dictionary<string, ReadOnlyMemory<byte>>
-                    {
-                        ["json"] = new byte[] {6, 7, 1, 2, 3, 4, 5},
-                        ["messagepack"] = new byte[] {7, 1, 2, 3, 4, 5, 6}
-                    }, tracingId: 1234L),
-                binary: "lQ2mZ3JvdXAzkIKkanNvbsQHBgcBAgMEBattZXNzYWdlcGFja8QHBwECAwQFBoEBzQTS"),
```
Comment (vwxyzh) [inline test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs:508]: why remove?

File: test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs [hunk 3]
```diff
@@ -402,7 +430,7 @@ public static IEnumerable<object[]> TestParseOldData
                         ["json"] = new byte[] {6, 7, 1, 2, 3, 4, 5},
                         ["messagepack"] = new byte[] {7, 1, 2, 3, 4, 5, 6}
                     }),
-                binary: "lQ2mZ3JvdXAzkIKkanNvbsQHBgcBAgMEBattZXNzYWdlcGFja8QHBwECAwQFBoA="),
+                binary: "lw2mZ3JvdXAzkIKkanNvbsQHBgcBAgMEBattZXNzYWdlcGFja8QHBwECAwQFBoCQwA=="),
```
Comment (vwxyzh) [inline test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs:433]: add the old one to `TestParseOldData`

Comment (vicancy) [reply test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs:433]: Already added to L253

File: src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs [hunk 4]
```diff
@@ -141,6 +141,16 @@ public class GroupBroadcastDataMessage : MulticastDataMessage
         /// </summary>
         public IReadOnlyList<string> ExcludedList { get; set; }
 
+        /// <summary>
+        /// Gets or sets the list of excluded user Ids.
+        /// </summary>
+        public IReadOnlyList<string> ExcludedUserList { get; set; }
```
Comment (vwxyzh) [inline src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:147]: should we support exclude connection id and user id at the same time?
if not, should we add an ExcludeTarget (connection id/user id) instead?

File: test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs (hunk 2 above)
Comment (vicancy) [reply test/Microsoft.Azure.SignalR.Protocols.Tests/ServiceProtocolFacts.cs:508]: reorder to merge with other `GroupBroadcastDataMessage` messages

File: src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs (hunk 1 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Protocols/ServiceProtocol.cs:874]: nice catch

File: src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs (hunk 4 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:147]: I think we can support both at the same time

File: src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs [hunk 5]
```diff
@@ -141,6 +141,16 @@ public class GroupBroadcastDataMessage : MulticastDataMessage
...
+
+        /// <summary>
+        /// Gets or sets the user ID of the sender
+        /// </summary>
+        public string SenderId { get; set; }
```
Comment (JialinXin) [inline src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:152]: CallerId?

**Review by vwxyzh — APPROVED** [review]

File: src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs (hunk 5 above)
Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:152]: Since it is used in `SendToGroup`, does `sender` make more sense?

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:152]: caller sounds more general and can be applied to other messages

Comment (vicancy) [reply src/Microsoft.Azure.SignalR.Protocols/MulticastDataMessage.cs:152]: since this is a `message`, I would prefer using `sender`.

**Review by JialinXin — APPROVED** [review]

**Pull request merged**